### ISSUE TITLE
Mmr checkpoints

### ISSUE THREAD

**Issue by hackaugusto** [0xMiden/miden-node#179]
The Mmr is a data structure that grows very slowly, and in addition to a slow growth, the Mmr has $log_2$ proof sizes, meaning the inclusion proofs are also very small:

| Period | Blocks | Bytes | MB | proof size (worst case) |
| ------ | ------: | -----: | --: | -----------------------: |
| day | 17280 | 2211584B | 2.11MB | 960B |
| week1 | 120960 | 15482496B | 14.77MB | 1088B |
| week2 | 241920 | 30965376B | 29.53MB | 1152B |
| week3 | 362880 | 46448256B | 44.30MB | 1216B |
| week4 | 483840 | 61931136B | 59.06MB | 1216B |
| month1 | 518400 | 66354688B | 63.28MB | 1216B |
| month2 | 1036800 | 132709888B | 126.56MB | 1280B |
| month3 | 1555200 | 199064960B | 189.84MB | 1344B |
| month4 | 2073600 | 265420288B | 253.12MB | 1344B |
| year1 | 6307200 | 807321088B | 769.92MB | 1472B |
| year2 | 12614400 | 1614642688B | 1539.84MB | 1536B |
| year3 | 18921600 | 2421964352B | 2309.77MB | 1600B |
| year4 | 25228800 | 3229285888B | 3079.69MB | 1600B |

(above 5sec block time was assumed)

<details>

<summary>Code used to produce above table</summary>

```rust
const BLOCK_TIME: usize = 5; // in seconds
const BLOCKS_DAY: usize = (24 * 60 * 60) / BLOCK_TIME;
const BLOCKS_WEEK: usize = BLOCKS_DAY * 7;
const BLOCKS_MONTH: usize = BLOCKS_DAY * 30;
const BLOCKS_YEAR: usize = BLOCKS_DAY * 365;
const BYTES_IN_MEGA: f64 = 1024_usize.pow(2) as f64;
const HASH_SIZE: usize = 64; // in bytes

fn size(forest: usize) -> usize {
    let tree_count = forest.count_ones() as usize;
    (forest * 2 - tree_count) * HASH_SIZE
}

fn biggest_tree(forest: usize) -> usize {
    ((usize::BITS - forest.leading_zeros()) as usize) * HASH_SIZE
}

fn main() {
    println!("| Period | Blocks | Bytes | MB | proof size (worst case) |");
    println!("| -----: | -----: | ----: | -: | ----------------------: |");

    println!(
        "| day | {} | {}B | {:.2}MB | {}B |",
        BLOCKS_DAY,
        size(BLOCKS_DAY),
        (size(BLOCKS_DAY) as f64) / BYTES_IN_MEGA,
        biggest_tree(BLOCKS_DAY),
    );
    for week in 1..=4 {
        println!(
            "| week{} | {} | {}B | {:.2}MB | {}B |",
            week,
            BLOCKS_WEEK * week,
            size(BLOCKS_WEEK * week),
            (size(BLOCKS_WEEK * week) as f64) / BYTES_IN_MEGA,
            biggest_tree(BLOCKS_WEEK * week),
        );
    }
    for month in 1..=4 {
        println!(
            "| month{} | {} | {}B | {:.2}MB | {}B |",
            month,
            BLOCKS_MONTH * month,
            size(BLOCKS_MONTH * month),
            (size(BLOCKS_MONTH * month) as f64) / BYTES_IN_MEGA,
            biggest_tree(BLOCKS_MONTH * month),
        );
    }
    for year in 1..=4 {
        println!(
            "| year{} | {} | {}B | {:.2}MB | {}B |",
            year,
            BLOCKS_YEAR * year,
            size(BLOCKS_YEAR * year),
            (size(BLOCKS_YEAR * year) as f64) / BYTES_IN_MEGA,
            biggest_tree(BLOCKS_YEAR * year),
        );
    }
}
```

</details>

Even if the `25228800` blocks created on the first 4 years, all would take `1600B`/`1.6KB` for inclusion proofs that would take only `37.59GB` of storage.

Given the above, it may be beneficial to store with each block its inclusion proof in the DB. This removes the need to synchronize the reading/writing of the Mmr, and streamlines the state_sync API. I think the only thing we need, is to ensure the client has some Mmr checkpoints, were say, every 50k blocks or so, the state sync api returns all the peaks to the client, and the db store the delta against those checkpoints.

**Comment by bobbinth** [comment]
> Given the above, it may be beneficial to store with each block its inclusion proof in the DB. This removes the need to synchronize the reading/writing of the Mmr, and streamlines the state_sync API. I think the only thing we need, is to ensure the client has some Mmr checkpoints, were say, every 50k blocks or so, the state sync api returns all the peaks to the client, and the db store the delta against those checkpoints.

Wouldn't this mean that we'd need to update each block's inclusion proof any time we merge the peaks? Or are you thinking of some other mechanism to store the proofs?

**Comment by hackaugusto** [comment]
> Wouldn't this mean that we'd need to update each block's inclusion proof any time we merge the peaks? Or are you thinking of some other mechanism to store the proofs?

I was thinking on making checkpoints/epochs, lets say the length of the epoch is `l`. Each of those would have a first block, call it `n`, the idea is to make all proofs of an epoch against said block `n`. Then:

- If there is nothing of interest to the client for the whole epoch, the server replies with the delta of the last block in the epoch, block `n + l`. This updates the client's partial mmr to the next epoch.
- If there is something of interest to the client, the server return the block with the inclusion proof w.r.t. the first block `n`. Which brings the client's Mmr to the block of interested.
     - If there are multiple blocks in the epoch that have data of interest to the client, then the client just has to discard the pieces of the diff that it has applied already. I think as long as we send the updates in order, this should work. Basically we send more data than the client needs.

With the above the writer can just insert the data to the DB, without synchronizing with the in-memory structure (because the readers only get data from the DB).

(I'm trying to figure out if it is possible to just define the thing in such a way we need a single scan of the DB)

**Comment by bobbinth** [comment]
I think MMR is one of the more compact data structures we have and keeping it in memory shouldn't be an issue.

For example, assuming 5-second block times, we get roughly 6M blocks per year. This should roughly translate to MMR size of less than 400MB. So, in 10 years, the size of the MMR would be less than 4GB. This is not really a problem even now, and I'm assuming that in 10 years this would be almost negligible amount of RAM on most computers.

**Comment by hackaugusto** [comment]
Yes, agreed. I wrote a very similar note above it on the original issue text. The goal here was not to save memory.
